src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer.UnitTests/Utilities/AbstractLanguageServerHostTests.TestLspServer.cs

```diff
@@ -0,0 +1,153 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.LanguageServer.LanguageServer;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Composition;
+using Nerdbank.Streams;
+using Roslyn.LanguageServer.Protocol;
+using Roslyn.Utilities;
+using StreamJsonRpc;
+using LSP = Roslyn.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.UnitTests;
+
+public partial class AbstractLanguageServerHostTests
```

Comment:
Instead of this, can we either
1.  Reuse the same testlsp server bits from the protocol unit tests (no need to test program startup / project system impl for these code action tests)
2.  Move the code action tests to the protocol project, but add a new MEF composition that includes the LanguageServer.dll (e.g. here - https://github.com/dotnet/roslyn/blob/main/src/LanguageServer/Protocol.TestUtilities/LspTestCompositions.cs#L11)

Reply:
Well the protocol unit tests are net472 and the LanguageServer is net8.0 so that is a bit of a non-starter. I like the idea you had to simply starting an instance of the LanguageServer and connecting to it as a client. We can shoot for that in a follow up or now if you think this is brittle.

Reply:
Updated to the full launch case.